src/org/rascalmpl/interpreter/env/ModuleEnvironment.java

```diff
@@ -327,15 +342,25 @@ public boolean isModuleEnvironment() {
 	
 	public void addImport(String name, ModuleEnvironment env) {
 		assert heap.getModule(name).equals(env);
-		importedModules.add(name);
+		importedModules.put(name, Optional.ofNullable(env));
 		typeStore.importStore(env.typeStore);
+		this.cachedGeneralKeywordParameters = null;
+		this.cachedPublicFunctions = null;
+	}
+
+	void removeModule(String name) {
+		importedModules.computeIfPresent(name, (k, v) -> Optional.empty());
```

Comment:
This requires some comments for future maintainers. We need to keep the import graph but remove the environment state. Otherwise they'll just replace by `.remove`

Reply:
I've put a javadoc on the `importedModules` field to explain this concern, do you prefer I repeat this everywhere?